File: 0876_middle_of_linked_list/middle_node.cpp
#include <iostream>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

/**
 * Time: O(n)
 * Space: O(1)
 */
ListNode *middle_node(ListNode *head)
{
    ListNode *h = head, *t = head;
    while (h && h->next)
    {
        h = h->next->next;
        t = t->next;
    }
    return t;
}

File: 0876_middle_of_linked_list/middle_node.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def middle_node(head: ListNode) -> ListNode:
    """
    Time: O(n)
    Space: O(1)
    """
    t = h = head
    while h and h.next:
        h = h.next.next
        t = t.next
    return t
